File: cpp/src/fileScanner.cpp
/*
 * Author: wilbur
 * Version: 1.0
 * Date: 2026-05-29
 * Description: 扫描顶层 JPG/RW2/CR2，按 stem 配对，不递归
 */

#include "fileScanner.h"
#include <filesystem>
#include <map>
#include <algorithm>
#include <stdexcept>

static bool endsWithIgnoreCase(const std::string& str, const std::string& suffix) {
    if (str.size() < suffix.size()) return false;
    for (size_t i = 0; i < suffix.size(); ++i) {
        if (std::tolower(str[str.size() - suffix.size() + i]) != std::tolower(suffix[i])) {
            return false;
        }
    }
    return true;
}

static bool isJpg(const std::string& filename) {
    return endsWithIgnoreCase(filename, ".jpg") || endsWithIgnoreCase(filename, ".jpeg");
}

static bool isRaw(const std::string& filename) {
    return endsWithIgnoreCase(filename, ".rw2") || endsWithIgnoreCase(filename, ".cr2");
}

std::vector<PhotoPair> FileScanner::scanTopLevel(const std::string& folderPath) const {
    namespace fs = std::filesystem;
    fs::path dir(folderPath);
    if (!fs::exists(dir) || !fs::is_directory(dir)) {
        throw std::runtime_error("Not a directory: " + folderPath);
    }

    std::map<std::string, PhotoPair> pairs;

    for (const auto& entry : fs::directory_iterator(dir)) {
        if (!entry.is_regular_file()) continue;
        std::string filename = entry.path().filename().string();
        std::string stem = entry.path().stem().string();
        std::string fullPath = entry.path().string();

        if (isJpg(filename)) {
            pairs[stem].photoId = stem;
            pairs[stem].jpgPath = fullPath;
            pairs[stem].hasJpg = true;
        } else if (isRaw(filename)) {
            pairs[stem].photoId = stem;
            pairs[stem].rawPath = fullPath;
            pairs[stem].hasRaw = true;
        }
    }

    std::vector<PhotoPair> result;
    for (auto& [stem, pair] : pairs) {
        result.push_back(pair);
    }

    std::sort(result.begin(), result.end(), [](const PhotoPair& a, const PhotoPair& b) {
        return a.photoId < b.photoId;
    });

    return result;
}

File: rawViewer/services/fileScanner.swift
/*
Author: wilbur
Version: 1.1
Date: 2026-06-13
Description: 顶层目录扫描, 按 stem 配对 JPG (jpg/jpeg) 和 RAW (rw2/cr2)。v1.1 明确扫描结果和扫描器可在后台分析任务中使用
*/

import Foundation

nonisolated public struct photoFilePair: Sendable {
    public let photoId: String
    public let jpgPath: String?
    public let rawPath: String?

    public init(photoId: String, jpgPath: String?, rawPath: String?) {
        self.photoId = photoId
        self.jpgPath = jpgPath
        self.rawPath = rawPath
    }

    public var hasJpg: Bool { jpgPath != nil }
    public var hasRaw: Bool { rawPath != nil }
}

nonisolated public final class fileScanner: @unchecked Sendable {
    private static let jpgExtensions: Set<String> = ["jpg", "jpeg"]
    private static let rawExtensions: Set<String> = ["rw2", "cr2"]

    public init() {}

    public func scanTopLevel(_ folderUrl: URL) throws -> [photoFilePair] {
        let fm = FileManager.default
        var isDir: ObjCBool = false
        guard fm.fileExists(atPath: folderUrl.path, isDirectory: &isDir), isDir.boolValue else {
            throw NSError(
                domain: "rawViewer.fileScanner", code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Not a directory: \(folderUrl.path)"]
            )
        }

        let items = try fm.contentsOfDirectory(at: folderUrl, includingPropertiesForKeys: nil, options: [.skipsHiddenFiles])
        var pairs: [String: photoFilePair] = [:]

        for url in items {
            let stem = url.deletingPathExtension().lastPathComponent
            let ext = url.pathExtension.lowercased()

            if Self.jpgExtensions.contains(ext) {
                pairs[stem] = photoFilePair(photoId: stem, jpgPath: url.path, rawPath: pairs[stem]?.rawPath)
            } else if Self.rawExtensions.contains(ext) {
                pairs[stem] = photoFilePair(photoId: stem, jpgPath: pairs[stem]?.jpgPath, rawPath: url.path)
            }
        }

        return pairs.values.sorted { $0.photoId < $1.photoId }
    }
}
